### PR TITLE
feat: add Cloud dedicated management api

### PR DESCRIPTION
**Pull request by jamesbalcombe83**
## Proposed Changes

Support [database creation for InfluxDB Cloud Dedicated](https://docs.influxdata.com/influxdb/cloud-dedicated/admin/databases/create/?t=Management+API).

## Checklist

- [x] CHANGELOG.md updated
- [x] Rebased/mergeable
- [x] A test has been added if appropriate
- [x] Tests pass
- [x] Commit messages are [conventional](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Sign [CLA](https://www.influxdata.com/legal/cla/) (if not already signed)

### CODE REVIEW

File: influxdb3/management_decicated.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		NumberOfBuckets uint64 `json:"numberOfBuckets"`
+	}
+)
+
+func (t Tag) IsPartitionTemplate()        {}
+func (tb TagBucket) IsPartitionTemplate() {}
+
+// NewCloudDedicatedClient creates new DedicatedClient with given InfluxDB client.
+func NewCloudDedicatedClient(client *Client) *DedicatedClient {
+	return &DedicatedClient{client: client}
```
Comment (xorphitus) [inline influxdb3/management_decicated.go:81]: The default domain `console.influxdata.com` should be set here, or the example file should show that users must pass it.

File: influxdb3/management_decicated.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+)
+
+type (
+	// Dedicatedclient represents a client for InfluxDB Cloud Dedicated administration operations.
+	DedicatedClient struct {
```
Comment (xorphitus) [inline influxdb3/management_decicated.go:37]: They may be tedious, but are `CloudDedicatedClient` and `CloudDedicatedClientConfig` better in terms of clarity for users?

**Review by xorphitus — COMMENTED** [review]
I made some comments from my experience with #86.

File: influxdb3/management_decicated.go (hunk 1 above)
Comment (jamesbalcombe83) [reply influxdb3/management_decicated.go:81]: Thanks, I forgot to add this into the example

File: influxdb3/management_decicated.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	Database struct {
+		Name               string              `json:"name"`
+		MaxTables          uint64              `json:"maxTables"`          // default 500
+		MaxColumnsPerTable uint64              `json:"maxColumnsPerTable"` // default 250
+		RetentionPeriod    uint64              `json:"retentionPeriod"`    // nanoseconds default 0 is infinite
+		PartitionTemplate  []PartitionTemplate `json:"partitionTemplate"`  // Tag or TagBucket, limit is total of 7
+	}
+
+	PartitionTemplate interface {
+		IsPartitionTemplate()
```
[the remark this answers is not in the document]
Comment (jamesbalcombe83) [reply influxdb3/management_decicated.go]: Fixed in [b09f5c9](https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/b09f5c954752378cad8b32c6f93a403a781a83bb)

File: influxdb3/management_decicated.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	if db == nil {
+		return errors.New("database must not nil")
+	}
+
+	if d.client.config.Database == "" {
+		return errors.New("database name must not be empty")
+	}
+	db.Name = d.client.config.Database
+
+	if len(db.PartitionTemplate) > 7 {
```
[the remark this answers is not in the document]
Comment (jamesbalcombe83) [reply influxdb3/management_decicated.go]: Fixed in [b09f5c9](https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/b09f5c954752378cad8b32c6f93a403a781a83bb)

File: influxdb3/management_decicated.go (hunk 1 above)
Comment (jamesbalcombe83) [reply influxdb3/management_decicated.go]: Fixed in [f8c58b5](https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/f8c58b5d83b7b04e9f603d176105e549321c1030) and [53a8696](https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/53a8696dc89b716e0c16117b60cf7e2d69e45834)

File: influxdb3/management_decicated.go (hunk 2 above)
Comment (jamesbalcombe83) [reply influxdb3/management_decicated.go]: Fixed in [b09f5c9](https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/b09f5c954752378cad8b32c6f93a403a781a83bb)

File: influxdb3/management_decicated.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+	Database struct {
+		Name               string              `json:"name"`
+		MaxTables          uint64              `json:"maxTables"`          // default 500
+		MaxColumnsPerTable uint64              `json:"maxColumnsPerTable"` // default 250
+		RetentionPeriod    uint64              `json:"retentionPeriod"`    // nanoseconds default 0 is infinite
+		PartitionTemplate  []PartitionTemplate `json:"partitionTemplate"`  // Tag or TagBucket, limit is total of 7
+	}
+
+	PartitionTemplate interface {
```
[the remark this answers is not in the document]
Comment (bednar) [reply influxdb3/management_decicated.go:56]: Agreed. What do you think about using the naming conventions from https://github.com/influxdata/docs-v2/blob/master/api-docs/cloud-dedicated/management/openapi.yml?

**Review by bednar — CHANGES_REQUESTED** [review]
@jamesbalcombe83 thanks a lot of for your PR 👍 

There is a failing CI check - https://app.circleci.com/pipelines/github/InfluxCommunity/influxdb3-go/307/workflows/5f5e1427-35cb-4c74-a09d-738afff129b5/jobs/317?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary

Can you take a look please?

**Comment by jamesbalcombe83** [comment]
> @jamesbalcombe83 thanks a lot of for your PR 👍
> 
> There is a failing CI check - https://app.circleci.com/pipelines/github/InfluxCommunity/influxdb3-go/307/workflows/5f5e1427-35cb-4c74-a09d-738afff129b5/jobs/317?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary
> 
> Can you take a look please?

@bednar I looked at this when the CI first failed, but what it's reporting isn't correct as far as I can tell. 
```
influxdb3/example_management_dedicated_test.go:32:1: ExampleDedicatedClient_CreateDatabase refers to unknown identifier: DedicatedClient
```
I checked the entire repo, and there is no such reference, I replaced them all with `CloudDedicatedClient`.

File: influxdb3/management_decicated.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jamesbalcombe83) [reply influxdb3/management_decicated.go:56]: @bednar @xorphitus As in adding `ClusterDatabase` to each of the parameters. This seems sensible to me.

[the remark this answers is not in the document]
Comment (jamesbalcombe83) [reply influxdb3/management_decicated.go]: I will go ahead with this change.

[the remark this answers is not in the document]
Comment (jamesbalcombe83) [reply influxdb3/management_decicated.go]: Addressed in (c4a4c2b)[https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/c4a4c2b0a589e9e3b4c50ffa0feb777a2c2acaff]

File: influxdb3/management_cloud_decicated.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+)
+
+type (
+	// CloudDedicatedClient represents a client for InfluxDB Cloud Dedicated administration operations.
```
Comment (bednar) [inline influxdb3/management_cloud_decicated.go:36]: Please add link to documentation and HTTP API.

File: influxdb3/client.go [hunk 7]
```diff
@@ -171,7 +171,7 @@ func (c *Client) makeAPICall(ctx context.Context, params httpParams) (*http.Resp
 		}
 	}
 	req.Header.Set("User-Agent", userAgent)
-	if c.authorization != "" {
+	if c.authorization != "" && req.Header.Get("Authorization") == "" {
```
Comment (bednar) [inline influxdb3/client.go:174]: Please add test to this change into `client_test.go`.

File: influxdb3/management_cloud_decicated.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	CloudDedicatedClient struct {
+		client *Client
+	}
+
+	CloudDedicatedClientConfig struct {
+		AccountID        string
+		ClusterID        string
+		ManagementToken  string
+		ManagementAPIURL *url.URL
```
Comment (bednar) [inline influxdb3/management_cloud_decicated.go:45]: Please add doc string what is used if the `ManagementAPIURL` is not specified.

**Review by bednar — CHANGES_REQUESTED** [review]
@jamesbalcombe83, thank you again for your PR 👍. There are a few requirements that need to be met before we can proceed with merging this PR:

File: influxdb3/management_cloud_decicated.go (hunk 6 above)
Comment (jamesbalcombe83) [reply influxdb3/management_cloud_decicated.go:36]: Addressed in [8a9191a](https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/8a9191ac852dc3770fbf39e7170889b872856df3)

File: influxdb3/management_cloud_decicated.go (hunk 8 above)
Comment (jamesbalcombe83) [reply influxdb3/management_cloud_decicated.go]: Addressed in [3abbb46](https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/3abbb46a86a5f2524087a5ec2e6ddf639faafcc2)

File: influxdb3/client.go (hunk 7 above)
Comment (jamesbalcombe83) [reply influxdb3/client.go:174]: Addressed in [d5b0fb9](https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/d5b0fb9cb5ba790990e6e383b255ddae1b4c4f42)

File: influxdb3/management_decicated.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (xorphitus) [reply influxdb3/management_decicated.go]: @bednar Perhaps, your force-push might removed the c4a4c2b0a589e9e3b4c50ffa0feb777a2c2acaff but I agree with @jamesbalcombe83 's prefixing idea.

[the remark this answers is not in the document]
Comment (bednar) [reply influxdb3/management_decicated.go]: It is just rebased with main. Is there something missing?

[the remark this answers is not in the document]
Comment (bednar) [reply influxdb3/management_decicated.go]: Check it out - https://github.com/InfluxCommunity/influxdb3-go/compare/d5b0fb9cb5ba790990e6e383b255ddae1b4c4f42..bde221f1a128bb86244dab8022452932535850d0

[the remark this answers is not in the document]
Comment (jamesbalcombe83) [reply influxdb3/management_decicated.go]: The commits I made to address your previous comments seem to have disappeared after the force-push

File: influxdb3/example_management_cloud_dedicated_test.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+		ClusterID:        clusterID,
+		ManagementToken:  managementToken,
+		ManagementAPIURL: managementAPIURL,
+	}
+
+	defer client.Close()
+
+	cloudDedicatedClient := NewCloudDedicatedClient(client)
+
+	if err := cloudDedicatedClient.CreateDatabase(context.Background(), &cloudDedicatedConfig, &Database{}); err != nil {
```
Comment (bednar) [inline influxdb3/example_management_cloud_dedicated_test.go:57]: Please add valid database structure
```suggestion
	database := &Database{
		.
		.
		.
	}
	if err := cloudDedicatedClient.CreateDatabase(context.Background(), &cloudDedicatedConfig, database); err != nil {
```

**Review by bednar — CHANGES_REQUESTED** [review]
@jamesbalcombe83, thank you for your PR 👍. There is one last requirement before we can merge the PR into the main branch:

File: influxdb3/management_decicated.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bednar) [reply influxdb3/management_decicated.go]: The changes are still here, this is a way how the GitHub rebases sources from `main` branch.

File: influxdb3/example_management_cloud_dedicated_test.go (hunk 9 above)
Comment (jamesbalcombe83) [reply influxdb3/example_management_cloud_dedicated_test.go]: @bednar Address in [80fb9eb](https://github.com/InfluxCommunity/influxdb3-go/pull/87/commits/80fb9ebe6bc90eee05300b0a83cebfe1dd11eb8d)

Comment (bednar) [reply influxdb3/example_management_cloud_dedicated_test.go]: 👍

**Review by bednar — APPROVED** [review]
LGTM 🚀

**Comment by jamesbalcombe83** [comment]
@bednar thanks for merging. Do you have a schedule for when this will be released?

**Comment by bednar** [comment]
@jamesbalcombe83, we release on a monthly basis, so the next release will likely be at the end of this month.

**Pull request merged**